### DISCUSSION TITLE
Customizing UI

### DISCUSSION THREAD

**Discussion by 4jag** [inventree/InvenTree#2820] [Q&A] [answered]
There are some basic UI layout changes that I'd like to make. I'm testing the production docker env (and having trouble on RHEL7 with pip3 and would like to avoid having to roll a dev server). Can basic changes in UI be implemented easily in prod/stable? 

For example, we have a LOT of locations for some parts, and the BOM view blows up due to the HTML layout giving nearly unlimited width to the data in the References field. I'd like to limit the width to ~150px and add a rollover popup with the full data.

**Comment by SchrodingersGat** [comment, marked as answer]
Making running changes to the code is difficult in the production environment. Any changes made to the python code would require a server restart.

However if you're only changing front-end web code (e.g. javascript / html) then you should be able to do this in the production environment.

You will need to edit the particular files within the running container, and then simply refreshing the browser should display the changes. 

Note that for javascript file changes, the javascript files are copied into the `static` directory, rather than being rendered on the fly. So you cannot edit the "source" javascript file, rather the copy which has been placed into the `static` directory.

Making any of these changes will be *much* simpler in a development environment. Perhaps you could setup a second environment for development and testing?

**Comment by 4jag** [comment]
Yes, I should stand up a Dev instance, but I'm on RHEL7 and having fits trying to install a working pip. :&

In a maven project that uses docker and keycloak I'm mapping in theme files using volumes, so this might be a good way to implement site-specific customizations. 
`runtime/src/main/container/filtered/runtime/docker-compose-keycloak.yml
services:
    keycloak:
        volumes:
            - ${PWD}/keycloak-config/themes/oursite:/opt/jboss/keycloak/themes/oursite`
            - 
I wonder if I can do the same here.

**Reply by matmair** [reply]
We have themes - those only cover css and are baked in. You can either fork and add your theme there or issue a 'condensed theme' in a PR.
There are currently no plans to add something more elaborate but we are always happy for PRs.
